GitUI/BranchTreePanel/RepoObjectsTree.Nodes.cs

```diff
@@ -177,56 +178,81 @@ public void StartAsyncRefreshTree()
             public bool IgnoreSelectionChangedEvent { get; set; }
             protected GitModule Module => UICommands.Module;
 
-            public void CancelReloadNodes()
+            protected bool IsAttached { get; private set; }
+
+            public Task AttachedAsync()
+            {
+                IsAttached = true;
+                return OnAttachedAsync();
+            }
+
+            protected virtual Task OnAttachedAsync()
+            {
+                return PostRepositoryChangedAsync();
+            }
+
+            public void Detached()
             {
                 _reloadCancellationTokenSequence.CancelCurrent();
+                IsAttached = false;
+                OnDetached();
+            }
+
+            protected virtual void OnDetached()
+            {
             }
 
             // Invoke from child class to reload nodes for the current Tree. Clears Nodes, invokes
             // input async function that should populate Nodes, then fills the tree view with its contents,
             // making sure to disable/enable the control.
-            protected async Task ReloadNodesAsync(Func<CancellationToken, Task<Nodes>> loadNodesTask, CancellationToken token)
+            protected async Task ReloadNodesAsync(Func<CancellationToken, Task<Nodes>> loadNodesTask)
             {
-                await TaskScheduler.Default;
-                token.ThrowIfCancellationRequested();
+                var token = _reloadCancellationTokenSequence.Next();
+
+                var treeView = TreeViewNode.TreeView;
+                if (treeView == null)
+                {
+                    return;
+                }
+
+                await treeView.SwitchToMainThreadAsync(token);
 
                 var newNodes = await loadNodesTask(token);
 
-                await TreeViewNode.TreeView.SwitchToMainThreadAsync(token);
-                token.ThrowIfCancellationRequested();
+                await treeView.SwitchToMainThreadAsync(token);
```

Comment:
❕ Just use `token.ThrowIfCancellationRequested()`. You are already on the UI thread here.